turbo/engineapi/request_list.go

```diff
@@ -154,7 +154,11 @@ func (rl *RequestList) updateWaiting(val uint32) {
 	if val == 1 {
 		// something might be waiting to be notified of the waiting state being ready
 		for _, c := range rl.waiters {
-			c <- struct{}{}
+			// ensure the channel is still open before writing to it
```

Comment:
Hmm, I did not actually found the place where these channels are closed, I was looking to see if closing happens under the same lock...

Reply:
channels are thread-safe. Mutex will not protect against sending to closed channel. Need wait until all producers/consumers shutdown the  close (or close by producer).

Reply:
I introduced the waiters here so can confirm they are only closed in the `stageLoopIsBusy` func in `ethbackend.go`.  This scenario could occur when the wait in `stageLoopIsBusy` times out after a second returning to the caller and the channel is closed before the loop over the waiters in `request_list.go` is triggered.  I didn't anticipate this originally and tests / hive were happy with the change.... until I ran up a node and spotted the panics.

This seems the cleanest solution.  Although I did consider returning a new channel from the call to `WaitForWaiting` so the channel creation and closing is the sole responsibility of the code in `request_list.go`.  The `stageLoopIsBusy` func could just wait on that channel being closed as a signal, so no code would actually be writing to the channel in the first place.   Any thoughts?

Reply:
"The stageLoopIsBusy func could just wait on that channel being closed as a signal, so no code would actually be writing to the channel in the first place" - now I understand general idea. 

My thoughts:
1. "create and close channel by producer" is general pattern to avoid "write to closed channel" problem (now you close channel by consumer). 
2. `WaitForWaiting` doesn't need aquier write mutex in the beginning of func, because mutex protect only `rl.waiters` field, all other logic don't need mutex. You can create `getter/setter` methods for `rl.waiters` field and only this methods will use mutex. Because "lock mutex and then write to channel" is bad pattern - because "write to channel" has undetermenistic time and "lock mutex for as-short-as-possible" is generally-good idea. 
3. `make(chan struct{})` for notifications better use buffered channels like `make(chan struct{}, 1)` then notifier can notify all subscribers without waiting for subscribers. Also when use buffered channels for notifications with 'at-least one' guranties, can  use non-blocking `select with default` because "if channel already full, then no difference - send there notification or drop it, subscribed will be notified 'at-least once' ".
4. Maybe `WaitForWaiting` can be implemented as loop with `syncCond.Wait`? Then it will be hard to time-out such wait. And actually this is your main problem: `syncCond.Wait` is not compatible with channels (with Select) in Go. So, maybe think about "not using syncCond".


Other:
5. "The stageLoopIsBusy func could just wait ..." - problem sounds similar to "send request to server, wait for response, but limit server-side work by timeout". In go we do solve this problem by `context.WithTimeout()`. I don't have clear advise here, just it sounds similar.

Reply:
1. Thanks for the tip, I always had in mind that the code that creates a channel should be responsible for closing it.  But the distinction between producer and consumer makes it clearer.  In this case should we just close the channel from the request_list side (producer) as we're just waiting on the signal, and it will always be to a single consumer?

2. Great idea, I'll get the implemented!

3. Great advice on using a buffered channel, never considered the select with default pattern there to avoid waiting.

4. I tried to use sync.Cond when first attempting this, and ended up with deadlocks as the call to Broadcast could happen before the call to Wait, which then stops any processing happening.  Also the timeout was really difficult to juggle as you mentioned.  I went to a channel based implementation after finding this.

5. I think normally this would make a lot of sense, just wait on the context.  But here we're just checking if some other process is busy or not, we haven't made a request to it, we're just checking its status, and want to make sure we wait a little while before confirming that it is ready or not.

Reply:
1. For example WaitForWaiting can return channel or channel + bool (for optimization - when don’t need wait).

Reply:
1. Or accept timeout as parameter and just block inside without returning channel.